src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs

```diff
@@ -2881,6 +2888,176 @@ public ConstructorEdit(INamedTypeSymbol oldType)
+                semanticEdits.Add(new SemanticEditInfo(SemanticEditKind.Update, symbolKey, syntaxMap, null, null));
+            }
+        }
+
+        private void ReportAttributeRudeEdits(ImmutableArray<AttributeData>? oldAttributes, ImmutableArray<AttributeData> newAttributes, ISymbol newSymbol, EditAndContinueCapabilities capabilities, ArrayBuilder<RudeEditDiagnostic> diagnostics, out bool needsEdit)
+        {
+            using var _ = ArrayBuilder<AttributeData>.GetInstance(out var changedAttributes);
+
+            FindChangedAttributes(oldAttributes, newAttributes, changedAttributes);
+            if (oldAttributes.HasValue)
+            {
+                FindChangedAttributes(newAttributes, oldAttributes.Value, changedAttributes);
+            }
+
+            needsEdit = false;
+            if (changedAttributes.Count == 0)
+            {
+                return;
+            }
+
+            // We need diagnostics reported if the runtime doesn't support changing attributes,
+            // but even if it does, only attributes stored in the CustomAttributes table are editable
+            if (!capabilities.HasFlag(EditAndContinueCapabilities.UpdateCustomAttributes) ||
+                changedAttributes.Any(IsNonCustomAttribute))
+            {
+                var newNode = FindSyntaxNode(newSymbol);
+                diagnostics.Add(new RudeEditDiagnostic(RudeEditKind.ChangingAttributesNotSupportedByRuntime, GetDiagnosticSpan(newNode, EditKind.Update), newNode, new[] { GetDisplayName(newNode, EditKind.Update) }));
+            }
+            else
+            {
+                needsEdit = true;
+            }
+
+            return;
+
+            static void FindChangedAttributes(ImmutableArray<AttributeData>? oldAttributes, ImmutableArray<AttributeData> newAttributes, ArrayBuilder<AttributeData> changedAttributes)
+            {
+                for (var i = 0; i < newAttributes.Length; i++)
+                {
+                    var newAttribute = newAttributes[i];
+                    var oldAttribute = FindMatch(newAttribute, oldAttributes);
+
+                    if (oldAttribute is null)
+                    {
+                        changedAttributes.Add(newAttribute);
+                    }
+                }
+            }
+
+            static AttributeData? FindMatch(AttributeData attribute, ImmutableArray<AttributeData>? oldAttributes)
+            {
+                if (!oldAttributes.HasValue)
+                {
+                    return null;
+                }
+
+                foreach (var match in oldAttributes.Value)
+                {
+                    if (SymbolEquivalenceComparer.Instance.Equals(match.AttributeClass, attribute.AttributeClass))
+                    {
+                        if (SymbolEquivalenceComparer.Instance.Equals(match.AttributeConstructor, attribute.AttributeConstructor) &&
+                            match.ConstructorArguments.SequenceEqual(attribute.ConstructorArguments) &&
```

Comment:
>match.ConstructorArguments.SequenceEqual(attribute.ConstructorArguments) [](http://example.com/codeflow?start=28&length=72)

This might not be sufficient for arguemnts that are arrays.
TypedConstant.Equals doesn't see to do deep equality comparison.

[https://sourceroslyn.io/#Microsoft.CodeAnalysis/Symbols/TypedConstant.cs,34](https://sourceroslyn.io/#Microsoft.CodeAnalysis/Symbols/TypedConstant.cs,34)